File: apps/extension/src/item-codec.ts
// Plaintext-shaped vault item content and its wire encoding as the JSON
// payload sealed inside an ItemRecord's ciphertext. Mirrors
// apps/web/src/vault/item-codec.ts field-for-field so items created by
// either client are readable by the other (the plaintext payload format
// is not a frozen contract file, but its wire shape must still match
// byte-for-byte with what actually gets sealed/opened) — apps/web is a
// forbidden path for this task, so this is a deliberate mirror, not a
// shared import, matching the existing account-bundle-codec.ts precedent
// of mirroring a shape across a language/runtime (here, a task-boundary)
// split. This module only ever touches plaintext bytes transiently, on
// the way into/out of the background's private CryptoWorkerHost.
export type ItemType = "login" | "note" | "totp-login";

export interface VaultItemData {
  type: ItemType;
  title: string;
  username?: string;
  password?: string;
  url?: string;
  notes?: string;
  /** Raw base32 TOTP secret, present only for type "totp-login". */
  totpSecret?: string;
}

export function encodeItemData(data: VaultItemData): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(data));
}

export function decodeItemData(bytes: Uint8Array): VaultItemData {
  const parsed = JSON.parse(new TextDecoder().decode(bytes));
  if (typeof parsed !== "object" || parsed === null || typeof parsed.title !== "string" || typeof parsed.type !== "string") {
    throw new Error("malformed vault item payload");
  }
  return parsed as VaultItemData;
}

/** Local input validation, mirroring apps/web's rules so the same item is
 * accepted or rejected identically by either client. */
export function validateItemData(data: VaultItemData): string[] {
  const problems: string[] = [];
  if (!data.title || data.title.trim().length === 0) problems.push("Title is required.");
  if (data.title && data.title.length > 500) problems.push("Title is too long.");
  if (data.type === "login" || data.type === "totp-login") {
    if (!data.username && !data.password) problems.push("A login needs a username or a password.");
  }
  if (data.type === "totp-login" && !data.totpSecret) {
    problems.push("A TOTP login needs a secret.");
  }
  if (data.notes && data.notes.length > 100_000) problems.push("Notes are too long.");
  return problems;
}

File: apps/android/app/src/main/kotlin/com/pass/android/codec/ItemCodec.kt
package com.pass.android.codec

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.nio.charset.StandardCharsets

/**
 * Port of `apps/extension/src/item-codec.ts` (itself a mirror of
 * `apps/web/src/vault/item-codec.ts`): plaintext-shaped vault item content
 * and its wire encoding as the JSON payload sealed inside an item's
 * ciphertext. Kept byte-for-byte compatible so items created by any client
 * are readable by the others. This module only ever touches plaintext
 * bytes transiently, in memory, on the way into/out of crypto-ffi.
 */
@Serializable
data class VaultItemData(
    val type: String, // "login" | "note" | "totp-login"
    val title: String,
    val username: String? = null,
    val password: String? = null,
    val url: String? = null,
    val notes: String? = null,
    /** Raw base32 TOTP secret, present only for type "totp-login". */
    val totpSecret: String? = null,
)

private val json = Json { encodeDefaults = false; ignoreUnknownKeys = true }

fun encodeItemData(data: VaultItemData): ByteArray =
    json.encodeToString(data).toByteArray(StandardCharsets.UTF_8)

class MalformedItemPayloadException : Exception("malformed vault item payload")

fun decodeItemData(bytes: ByteArray): VaultItemData {
    val text = String(bytes, StandardCharsets.UTF_8)
    val parsed = try {
        json.decodeFromString<VaultItemData>(text)
    } catch (e: Exception) {
        throw MalformedItemPayloadException()
    }
    if (parsed.title.isEmpty() && parsed.type.isEmpty()) throw MalformedItemPayloadException()
    return parsed
}

/** Local input validation, mirroring apps/web/apps/extension's rules so the
 * same item is accepted or rejected identically by every client. */
fun validateItemData(data: VaultItemData): List<String> {
    val problems = mutableListOf<String>()
    if (data.title.isBlank()) problems.add("Title is required.")
    if (data.title.length > 500) problems.add("Title is too long.")
    if (data.type == "login" || data.type == "totp-login") {
        if (data.username.isNullOrEmpty() && data.password.isNullOrEmpty()) {
            problems.add("A login needs a username or a password.")
        }
    }
    if (data.type == "totp-login" && data.totpSecret.isNullOrEmpty()) {
        problems.add("A TOTP login needs a secret.")
    }
    if ((data.notes?.length ?: 0) > 100_000) problems.add("Notes are too long.")
    return problems
}
